File: crates/fc-sdk/src/client/audit_logs.rs
//! Audit log query operations.

use serde::{Deserialize, Serialize};
use super::{FlowCatalystClient, ClientError, ListResponse};

/// Filters for listing audit logs.
#[derive(Debug, Clone, Default)]
pub struct AuditLogFilters {
    pub entity_type: Option<String>,
    pub entity_id: Option<String>,
    pub operation: Option<String>,
    pub principal_id: Option<String>,
    pub client_id: Option<String>,
    pub from: Option<String>,
    pub to: Option<String>,
    pub page: Option<u32>,
    pub page_size: Option<u32>,
}

/// Audit log entry from the platform API.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AuditLogResponse {
    pub id: String,
    pub operation: String,
    pub entity_type: String,
    #[serde(default)]
    pub entity_id: Option<String>,
    #[serde(default)]
    pub principal_id: Option<String>,
    #[serde(default)]
    pub principal_name: Option<String>,
    #[serde(default)]
    pub application_id: Option<String>,
    #[serde(default)]
    pub client_id: Option<String>,
    pub performed_at: String,
}

impl FlowCatalystClient {
    /// List audit logs with optional filters.
    pub async fn list_audit_logs(
        &self,
        filters: &AuditLogFilters,
    ) -> Result<ListResponse<AuditLogResponse>, ClientError> {
        let mut params = Vec::new();
        if let Some(ref v) = filters.entity_type {
            params.push(format!("entityType={}", v));
        }
        if let Some(ref v) = filters.entity_id {
            params.push(format!("entityId={}", v));
        }
        if let Some(ref v) = filters.operation {
            params.push(format!("operation={}", v));
        }
        if let Some(ref v) = filters.principal_id {
            params.push(format!("principalId={}", v));
        }
        if let Some(ref v) = filters.client_id {
            params.push(format!("clientId={}", v));
        }
        if let Some(ref v) = filters.from {
            params.push(format!("from={}", v));
        }
        if let Some(ref v) = filters.to {
            params.push(format!("to={}", v));
        }
        if let Some(v) = filters.page {
            params.push(format!("page={}", v));
        }
        if let Some(v) = filters.page_size {
            params.push(format!("pageSize={}", v));
        }
        let query = if params.is_empty() {
            String::new()
        } else {
            format!("?{}", params.join("&"))
        };
        self.get(&format!("/api/audit-logs{}", query)).await
    }

    /// Get a single audit log entry by ID.
    pub async fn get_audit_log(&self, id: &str) -> Result<AuditLogResponse, ClientError> {
        self.get(&format!("/api/audit-logs/{}", id)).await
    }
}

File: frontend/src/api/audit-logs.ts
/**
 * API client for Audit Log operations.
 */

import { apiFetch } from "./client";
import type {
	AuditLogApplicationIdsResponse,
	AuditLogClientIdsResponse,
	AuditLogEntityTypesResponse,
	AuditLogListResponse as GenAuditLogListResponse,
	AuditLogOperationsResponse,
	AuditLogResponse,
} from "./generated";

// Response types alias the generated contract (api/openapi.lock.json) so
// `vue-tsc` fails on backend drift. Aliased under the historical names so
// pages keep their imports. Absent fields are `undefined` (not `null`) on
// the wire, and the detail endpoint returns the same AuditLogResponse —
// `operationJson` lives on the base shape (optional).
export type AuditLog = AuditLogResponse;
export type AuditLogDetail = AuditLogResponse;

/**
 * Cursor-paginated audit logs response. Backend keysets on
 * `(performedAt, id) DESC` and never counts — `aud_logs` is unbounded.
 */
export type AuditLogListResponse = GenAuditLogListResponse;

export interface AuditLogFilters {
	entityType?: string;
	entityId?: string;
	principalId?: string;
	operation?: string;
	applicationIds?: string[];
	clientIds?: string[];
	/** Opaque cursor returned by a previous response. Omit for the first page. */
	after?: string | undefined;
	pageSize?: number;
}

/**
 * Fetch a page of audit logs (cursor-paginated).
 */
export async function fetchAuditLogs(
	filters: AuditLogFilters = {},
): Promise<AuditLogListResponse> {
	const params = new URLSearchParams();
	if (filters.entityType) params.set("entityType", filters.entityType);
	if (filters.entityId) params.set("entityId", filters.entityId);
	if (filters.principalId) params.set("principalId", filters.principalId);
	if (filters.operation) params.set("operation", filters.operation);
	if (filters.applicationIds?.length) params.set("applicationIds", filters.applicationIds.join(","));
	if (filters.clientIds?.length) params.set("clientIds", filters.clientIds.join(","));
	if (filters.after) params.set("after", filters.after);
	if (filters.pageSize !== undefined)
		params.set("pageSize", String(filters.pageSize));

	const query = params.toString();
	return apiFetch<AuditLogListResponse>(
		`/audit-logs${query ? `?${query}` : ""}`,
	);
}

/**
 * Fetch a single audit log by ID.
 */
export async function fetchAuditLogById(id: string): Promise<AuditLogDetail> {
	return apiFetch<AuditLogDetail>(`/audit-logs/${id}`);
}

/**
 * Fetch audit logs for a specific entity.
 */
export async function fetchAuditLogsForEntity(
	entityType: string,
	entityId: string,
): Promise<AuditLogListResponse> {
	return apiFetch<AuditLogListResponse>(
		`/audit-logs/entity/${encodeURIComponent(entityType)}/${encodeURIComponent(entityId)}`,
	);
}

/**
 * Fetch distinct entity types that have audit logs.
 */
export async function fetchEntityTypes(): Promise<AuditLogEntityTypesResponse> {
	return apiFetch<AuditLogEntityTypesResponse>("/audit-logs/entity-types");
}

/**
 * Fetch distinct operations that have audit logs.
 */
export async function fetchOperations(): Promise<AuditLogOperationsResponse> {
	return apiFetch<AuditLogOperationsResponse>("/audit-logs/operations");
}

/**
 * Fetch distinct application IDs present in audit logs.
 */
export async function fetchDistinctApplicationIds(): Promise<AuditLogApplicationIdsResponse> {
	return apiFetch<AuditLogApplicationIdsResponse>("/audit-logs/application-ids");
}

/**
 * Fetch distinct client IDs present in audit logs.
 */
export async function fetchDistinctClientIds(): Promise<AuditLogClientIdsResponse> {
	return apiFetch<AuditLogClientIdsResponse>("/audit-logs/client-ids");
}
